tools/e2etesting/TestEventProcessor/TestEventProcessor.Businesslogic/EventProcessorWrapper.cs

```diff
@@ -0,0 +1,254 @@
+        }
+
+        /// <summary>
+        /// Initialize EventProcessorClient.
+        /// </summary>
+        public async Task InitializeClient(CancellationToken ct) {
+            if (_client != null) {
+                return;
+            }
+
+            // Initialize EventProcessorClient
+            _logger.LogInformation("Connecting to blob storage...");
+            var blobContainerClient = new BlobContainerClient(
+                _config.StorageConnectionString,
+                _config.BlobContainerName
+            );
+
+            // Get number of partitions and initialize _initializedPartitions.
+            var eventHubConsumerClient = new EventHubConsumerClient(
+                _config.EventHubConsumerGroup,
+                _config.IoTHubEventHubEndpointConnectionString
+            );
+
+            var partitions = await eventHubConsumerClient
+                .GetPartitionIdsAsync(ct)
+                .ConfigureAwait(false);
+            _initializedPartitions = partitions.ToDictionary(item => item, _ => false);
+            _lockInitializedPartitions = new SemaphoreSlim(1, 1);
+
+            _logger.LogInformation("Connecting to IoT Hub...");
+            _client = new EventProcessorClient(
+                blobContainerClient,
+                _config.EventHubConsumerGroup,
+                _config.IoTHubEventHubEndpointConnectionString
+            );
+
+            _client.PartitionClosingAsync += Client_PartitionClosingAsync;
+            _client.PartitionInitializingAsync += Client_PartitionInitializingAsync;
+            _client.ProcessEventAsync += Client_ProcessEventAsync;
+            _client.ProcessErrorAsync += Client_ProcessErrorAsync;
+        }
+
+        /// <summary>
+        /// Start processing of events. If the method was called previously, then it will only
+        /// re-enable processing.
+        /// </summary>
+        public async Task StartProcessingAsync(CancellationToken ct) {
+            if (_client == null) {
+                throw new InvalidOperationException("EventProcessorWrapper has not been initialized.");
+            }
+
+            // If processing is active for all partitions then we only need to enable message propagation.
+            _lockInitializedPartitions.Wait();
+            try {
+                if (_initializedPartitions.Count(kvp => kvp.Value) == _initializedPartitions.Count) {
+                    _logger.LogInformation("Enabling monitoring of events...");
+                    _monitoringEnabled = true;
+                    return;
+                }
+            }
+            finally {
+                _lockInitializedPartitions.Release();
+            }
+
+            if (!_client.IsRunning) {
+                _logger.LogInformation("Starting monitoring of events...");
+                await _client.StartProcessingAsync(ct).ConfigureAwait(false);
+            }
+
+            await WaitForPartitionInitialization(ct).ConfigureAwait(false);
```

Comment:
>WaitForPartitionInitialization [](http://example.com/codeflow?start=18&length=30)

shouldn't we have this within the _lockInitializedPartitions block?

Reply:
No, `WaitForPartitionInitialization()` takes care of locking itself. It actually needs to release the lock in between the checks so that `Client_PartitionInitializingAsync()` can update the dictionary.